### PR TITLE
Allow visibility on constructors

### LINKED ISSUE #2341: Allow visibility on constructors
I think it is a pretty common pattern to have a static factory method to create a class and the constructor of this class being otherwise private so that you cannot instantiate the class unless you use the factory method.

### PR DESCRIPTION
**Pull request by AbubakerB**
Fixes #2341.

Changes:
1 - Allows private, protected and public accessibilities on constructors.
2 - Disallows classes with private constructors from being extended.
3 - Now throws an error on constructor overloads with different access modifiers (than the constructors implementation).

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
For this to work properly, the visibility needs to be part of the construct signature rather than being a declaration-based check that happens during signature resolution. Otherwise it's going to be too difficult to enforce assignability constraints based on constructor visibility.

A relevant testcase to add would be:
```ts
class Foo {
  private constructor() { }
}
class Bar {
  protected constructor() { }
}
let x: new() => any;
x = Foo; // Should be an error
x = Bar; // Should be an error
```

**Comment by AbubakerB** [comment]
@RyanCavanaugh SignaturesRelatedTo now checks visibility of constructor.

**Comment by AbubakerB** [comment]
Should i close this off? or is this in the pipeline?

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -8697,6 +8723,37 @@ namespace ts {
             }
 
             return resolveErrorCall(node);
+            
+            function checkConstructorVisibility(signature: Signature, node: NewExpression) {
+                let constructor = result.declaration;                
+                if (!constructor || !node.expression) return;
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Can you wrap these `return` statements with curlies and place them on their own line?

```TypeScript
if (!constructor || !node.expression) {
    return;
}
```

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -8697,6 +8723,37 @@ namespace ts {
             }
 
             return resolveErrorCall(node);
+            
+            function checkConstructorVisibility(signature: Signature, node: NewExpression) {
+                let constructor = result.declaration;                
+                if (!constructor || !node.expression) return;
+                // if constructor is public, we dont need to check visibility
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: dont -> don't

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -5097,6 +5097,13 @@ namespace ts {
                 let targetSig = targetSignatures[0];
 
                 if (sourceSig && targetSig) {
+                    if (kind === SignatureKind.Construct) { 
+                        result &= signatureVisibilityRelatedTo(sourceSig, targetSig);
+                        if (result !== Ternary.True) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Maybe I'm missing something shouldn't this be `===`?

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -15003,7 +15054,25 @@ namespace ts {
...
+                        // then the derived class should not be allowed to extend it.
+                        if (heritageClause.types.length == 1) {
+                            let expression = heritageClause.types[0].expression;
+                            if (expression) {
+                                let baseType = getApparentType(checkExpression(expression));
+                                if (baseType !== unknownType) {
+                                    let signatures = getSignaturesOfType(baseType, SignatureKind.Construct);
+                                    for (let signature of signatures) {
+                                        let constuctor = signature.declaration;
+                                        if (constuctor && constuctor.flags & NodeFlags.Private) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Wrap `constuctor.flags & NodeFlags.Private` with parentheses

**Comment by DanielRosenwasser** [comment]
The declaration emitter will need to be amended - add `// @declaration: true` to the top of every file.

**Comment by DanielRosenwasser** [comment]
The declaration emitter will need to be amended - add `// @declaration: true` to the top of every affected test file.

**Comment by DanielRosenwasser** [comment]
The following test cases would be useful too:

```TypeScript
class Base {
    protected constructor() {
    }
}

class Derived extends Base {
    private constructor() {
        super();
    }
}
```

```TypeScript
class Base {
    protected constructor() {
    }
}

class Derived extends Base {
    protected constructor() {
        super();
    }
}

var baseCtor = Base;
baseCtor = Derived;
```

**Comment by DanielRosenwasser** [comment]
@AbubakerB sorry this fell off of our radar, we've been hard at work on TypeScript 1.6.

@ahejlsberg @mhegazy @vladima any thoughts?

**Comment by AbubakerB** [comment]
@DanielRosenwasser I see, that's fine :)

1. I've added ```// @declaration: true``` to the three test files.
2. Added test cases under ```classConstructorAccessibility2.ts```. Although, I have one question... Can a X ctor signature be assigned to a non-X ctor signature? (where X is public/protected/private).
3. Renamed the diagnostic message for error code 2656 to be more clear.

File: src/compiler/checker.ts (hunk 1 above)
Comment (AbubakerB) [inline src/compiler/checker.ts]: Done

File: src/compiler/checker.ts (hunk 2 above)
Comment (AbubakerB) [inline src/compiler/checker.ts]: Done

File: src/compiler/checker.ts (hunk 4 above)
Comment (AbubakerB) [inline src/compiler/checker.ts]: Done

File: src/compiler/checker.ts (hunk 3 above)
Comment (AbubakerB) [inline src/compiler/checker.ts]: I thought I should follow the same structure as [here](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/checker.ts#L5087).

It checks to see if its not assignable (resulted from ```signatureVisibilityRelatedTo```) and returns if not.

File: tests/baselines/reference/classConstructorAccessibility.errors.txt [hunk 5]
```diff
@@ -1,29 +1,39 @@
-tests/cases/conformance/classes/constructorDeclarations/classConstructorAccessibility.ts(6,5): error TS1089: 'private' modifier cannot appear on a constructor declaration.
-tests/cases/conformance/classes/constructorDeclarations/classConstructorAccessibility.ts(10,5): error TS1089: 'protected' modifier cannot appear on a constructor declaration.
-tests/cases/conformance/classes/constructorDeclarations/classConstructorAccessibility.ts(23,9): error TS1089: 'private' modifier cannot appear on a constructor declaration.
-tests/cases/conformance/classes/constructorDeclarations/classConstructorAccessibility.ts(27,9): error TS1089: 'protected' modifier cannot appear on a constructor declaration.
+tests/cases/conformance/classes/constructorDeclarations/classConstructorAccessibility.ts(20,9): error TS2655: Constructor '(x: number): D' is private and only accessible within class 'D'.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/classConstructorAccessibility.errors.txt]: Shouldn't this be `new (x: number): D`?

File: src/compiler/diagnosticMessages.json [hunk 6]
```diff
@@ -1697,6 +1697,22 @@
...
     },
+    "Cannot extend private class '{0}'.": {
+        "category": "Error",
+        "code": 2653
+    },
+    "Constructor '{0}' is protected and only accessible within class '{1}' and its subclasses.": {
+        "category": "Error",
+        "code": 2654
+    },
+    "Constructor '{0}' is private and only accessible within class '{1}'.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: Maybe it should be "of type `'{0}'`

Comment (AbubakerB) [inline src/compiler/diagnosticMessages.json]: Done

File: tests/baselines/reference/classConstructorAccessibility.errors.txt (hunk 5 above)
Comment (AbubakerB) [inline tests/baselines/reference/classConstructorAccessibility.errors.txt]: What would be the best way of adding the ```new``` before the ```signature``` output?
I can't seem to find any method that writes ```'new' + signature```.

**Comment by RyanCavanaugh** [comment]
@AbubakerB can you get this up-to-date and resubmit in a new PR? Very sorry we left this hanging for so long.

**Comment by AbubakerB** [comment]
Sure. I've opened a new PR for this.

**Pull request closed without merging**